redpanda-data/redpanda
# pp: add basic auth to /brokers

## Cover letter

At Redpanda there is an effort to improve security by support HTTP Basic Authentication to our RESTful services (e.g., Pandaproxy and Schema Registry). The Pandaproxy will need to account for multiple authenticated connections on top of validating user credentials from the Basic Auth header.

This PR adds a function wrapper to all Pandaproxy route handlers where basic authentication support is within the wrapper. Inside the wrapper is where credentials are taken from the auth header and then validated against the shared credential store.

This PR adds support for multiple authenticated connections by creating a Kafka client cache. The cache is designed as a sharded service, it has a LRU replacement policy, and all clients are destroyed after 30s of liveness. The cache is added to the brokers route only in this patch.

The new configuration options to enable basic auth are attached to the listener in pandaproxy_api:
```yaml
authentication_method: object
  authentication_type: string
    Optional values: null, http_basic, mTLS
    Default value: null
```

Examples for the new authn config:
```yaml
pandaproxy:
  pandaproxy_api:
  - address: localhost
    port: 8082
    authentication_method:
      authentication_type: none
```

```yaml
pandaproxy:
  pandaproxy_api:
  - address: localhost
    port: 8082
    authentication_method:
      authentication_type: mtls
```

```yaml
pandaproxy:
  pandaproxy_api:
  - address: localhost
    port: 8082
    authentication_method:
      authentication_type: http_basic
```
After this PR, the leftover tasks are:
- Add support for multiple authenticated connections to the remaining Pandaproxy routes `/topic` and `/consumers`
- Add Basic Auth to Schema Registry routes `/config`, `/mode`, `/schemas`, `/subjects`, `/compatibility`
- Create upgrade, scale, and feature tests in ducktape
- Manual tests on cloud infra

Changes from force-push `e75e886` and `4c63f28`:
- Cosmetic changes
- Updated pre-existing auth wrapper to capture passwords
- Removed custom `rest_request_authenticator` class
- Placed authentication in a central location, `wrap()`
- Changed credential type to be a struct
- Capturing kafka api configuration in the client cache
- Update client cache to assign a principal when the Kafka subsystem has SASL enabled
- Check if the incoming request is for a connection with Basic Auth enabled

## Backport Required


- [x] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

* none

## Release notes

* Adds HTTP Basic Auth to /brokers on the Pandaproxy
